nats_test.go

```diff
@@ -1130,3 +1131,39 @@ func TestProcessErrAuthorizationError(t *testing.T) {
 		t.Fatal("Expected error on channel")
 	}
 }
+
+func TestConnAsyncCBDeadlock(t *testing.T) {
+	s := RunServerOnPort(DefaultPort)
+	defer s.Shutdown()
+
+	ch := make(chan bool)
+	o := GetDefaultOptions()
+	o.Url = DefaultURL
+	o.ClosedCB = func(_ *Conn) {
+		ch <- true
+	}
+	o.AsyncErrorCB = func(nc *Conn, sub *Subscription, err error) {
+		// do something with nc that requires locking behind the scenes
+		_ = nc.LastError()
+	}
+	nc, err := o.Connect()
+	if err != nil {
+		t.Fatalf("Should have connected ok: %v", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < 10000; i++ {
+		wg.Add(1)
+		go func() {
```

Comment:
nit but probably do not need 1k Go routines, just one that does the iteration loop to cause the channel to become full.

Reply:
Actually would not even work if running with `-race`, which we do, because there is a limit of 8192 go routines in that mode.